### ISSUE TITLE
Elias; Behavior Strategy; 3 WEEKS

### ISSUE THREAD

**Issue by EliasMlopez99** [lab-brussels-1/home#361]
# Learning objectives

### [1. Remix](./1-remix/)

Practice studying and remixing other people's solutions to coding challenges. Create your own solutions by mixing and matching pieces from other people's code.

- 🐣 **Reconstructing**: You can reconstruct a variety of solutions to the same challenge when they are presented as parsons problems.
- 🐣 **Analyzing**: You can analyze a function written at your level. This includes:
  - [ ] **Behavior**: Write documentation, test cases and use cases to describe the function's behavior.
  - [ ] **Strategy**: Describe the function's strategy using plain english.
  - [ ] **Implementation**: List the language features in a function and explain how each one is used.
  - [ ] **Small Changes**: You can think of 2+ changes to the function's _implementation_ that would not change it's _strategy_.
- 🐣 **Remixing** You can analyze several solutions to the same challenge then ...
  - [ ] **Write**: Your own solution by remixing the ones you studied.
  - [ ] **Explain**: How the other solutions inspired yours; what ideas did you take from them? what ideas did you _not_ take?
  - [ ] **Analyze**: Complete a write-up your own solution as though someone else wrote it.

### [2. Write](./2-write/)

- 🐣 **Function Design**:
  - [ ] **Writing Tests**: Given a working function, you can write passing test cases to describe it's behavior.
  - [ ] **Writing Functions**: You can design multiple solutions to the same code challenge, keeping notes about different experiments you tried along the way.
- [ ] 🐣 **Generating Documentation**: You can write a JSDoc comment for your solutions and run a script to generate markdown documentation.
- [ ] 🐣 **Fuzz Testing**: You write solutions that pass randomly generated test cases.
- 🐣 **Test Driven Development**: You can solve open-ended, ambiguous coding challenges at your level:
  - [ ] **Reading Docs**: You can understand what the function is supposed to do by reading it's JSDoc description.
  - [ ] **Writing Tests**: You can write test cases _before_ there is a function to test.
  - [ ] **Writing Functions**: You can write one function that passes the test cases you have prepared (even if it's just 1 test case!).
  - [ ] **Refactoring**: You can improve your function's implementation _without_ failing any test cases that were passing.
  - [ ] **Iterative Development**: You can repeat the TDD process until you are satisfied with your test cases and solution.
- [ ] 🐔 **Code Golf**: Write your solutions with the fewest characters possible! This won't help you write readable code, but it will make you think deeply about JS, your strategy and implementation.

### [3. Review](./3-review/)

- 🐣 **Continuous Integration**: You can check your code's quality _before_ pushing so your CI checks all pass.
  - [ ] **Formatting**
  - [ ] **Linting**
  - [ ] **Testing**
- [ ] 🐣 **Code Review**: You can use a checklist to give a thorough, positive and constructive review of your classmates' solutions.
- [ ] 🐔 **Code Coverage**: You can explain what code coverage is, why it's important, and can write unit tests with 100% code coverage.

**Comment by danielhalasz** [comment]
> * I have pushed my progress to [my fork of exercises repo](https://github.com/EliasMlopez99/behavior-strategy-implementation/blob/my-solution/1-remix/parsons-problems/add-arrays/map-method-2.js)
> 
> *

thanks for the links and the creative problemsolving :) the idea originally was that you fork this repo and write up 3 solutions there https://github.com/HackYourFutureBelgium/solution-write-ups

**Comment by danielhalasz** [comment]
> just the only thing that annoys me a bit, is closing my codification with )}] you have to be very precise where you place them

I feel the pain, this can be frustrating. Luckily there are some VSCode extensions that can help.

**Comment by danielhalasz** [comment]
> i want to say for the first time that i really had fun in the javascript class of the past Sunday :)

💟

**Comment by danielhalasz** [comment]
can you describe what was fun about it?

**Comment by danielhalasz** [comment]
Is there anything in the screenshot that you wanted to show? It seems to have some interesting lines..
like lines 49, 53, 62, 66... where you use < > & operators.. hope you have fixed that 🙂 and see why that is problematic this way.

**Comment by danielhalasz** [comment]
> I know theory is important, however sometimes "testing yourself" is way better than letting other to tell you how "it could be"

yes, this is also our goal, and that's why there are so many exercises and examples in all the modules and repos, not just this one, but the previous ones too 🙂

**Comment by danielhalasz** [comment]
> i would really like to discuss with one of the coaches of HYF if possible, like a small zoom, it would be really great if it was thibaut do not get me wrong the other coaches were just fine but i understood more js within this module, than in the past ones.

sure, we can ask Thibault, if he can be available for a followup session one day with the class. is there any specific question that you would ask him? what functions did you get stuck on?

**Comment by danielhalasz** [comment]
> Well as i mention above just some "tips" "advices" creating functions, i know i'm suppose to read how to do them but i consider i find it better if someone with experience show me how to.

can you please try to be more specific about the question? what do you mean by creating functions? declaring functions? function syntax? the scope? the content? parameters? it is indeed highly recommended to first look at the suggested study materials..if you dont like articles, there are plenty of videos online in multiple languages..this is how it will be in real life as well working at a company.. and when after putting in the effort, you still dont understand something, then of course you ask for help..

**Comment by danielhalasz** [comment]
> Well i wish we could have more codification sections instead of long 'blocks' of explanation, i know that this program has been like this for years though with this matters i prefer practical than the theory. Something like "check out this example" "test it" "try it" "see how it works" now it's your turn.

well this is exactly what we are doing. showing examples that are harder to grasp on Sunday, starting teamwork/individual work..and keep practicing during the week until the next Sunday session.. if we did even less explanation on Sundays, it would require you to prepare and learn on your own even more..which as you said above, might not be the best for everyone..so there has to be a balance.. it would be great if everyone came prepared with the knowledge and we could just do exercises in class.
